### ISSUE TITLE
HTTP Healthcheck responds with 200 despite no tailnet ip

### ISSUE THREAD

**Issue by chrismilson** [tailscale/tailscale#13620]
### What is the issue?

Steps to reproduce:
- Start a tailscale container and set the TS_HEALTHCHECK_ADDR_PORT environment variable.
- The container will register itself in the tailnet successfully using its token.
- The healthcheck will start to respond with HTTP 200 OK.
- Manually remove the device from the tailnet via the UI.

Expectation: Healthcheck for the container will start responding with HTTP 503.

Reality: Healthcheck continues to respond with HTTP 200 OK.

I noticed that `tailscale status` shows the container node as `"Online": false`.

### Are there any recent changes that introduced the issue?

#12898

### Tailscale version

v1.74.1

**Comment by chrismilson** [comment]
I am currently running tailscale in Kubernetes (without the Operator, more fool me), and in cases where the cluster has lost internet connectivity momentarily, the pods have become offline. Importantly, they did not come back online after their connectivity was restored. Ideally in this situation I would like the health checks to start failing so that they will be terminated and rescheduled, but your point about "offline but still healthy" is quite interesting.

I'm not so familiar with how tailscale keeps connections between peers while the control plane is down, but in theory: if a pod (tailscale node) lost connection to the control plane, I would want the liveness probe to fail immediately, so that I would be able to configure my own threshold for how many failed probes occur before the pod is terminated.  

There is probably a way of me doing this already by installing jq somewhere to correctly parse the json status in an `exec` probe.

Ideally though there would be a /healthz/online or something endpoint that I could configure on the `httpGet` probe that I could call instead.

**Comment by tomhjp** [comment]
Thanks for the extra detail @chrismilson!

> I'm not so familiar with how tailscale keeps connections between peers while the control plane is down

We have a [good article on this](https://tailscale.com/kb/1091/what-happens-if-the-coordination-server-is-down) if you're interested.

Separately, I'm curious about the report that pods didn't reconnect after getting internet connectivity back. I tried to reproduce this but failed:

```bash
export TS_AUTHKEY=...
docker run --name ts-test --hostname ts-test --rm -e TS_USERSPACE=false -e TS_AUTHKEY --device /dev/net/tun --cap-add NET_ADMIN tailscale/tailscale
```

In another terminal (with some sensible short pauses in between commands):

```bash
tailscale ping ts-test
# Success

docker network disconnect bridge ts-test
tailscale ping ts-test
# Failure

docker network connect bridge ts-test
tailscale ping ts-test
# Success
```

Are you able to quantify any important environmental details that differ in what you observed?

In your case you say connectivity was only disrupted momentarily, but one way I think I _could_ reproduce this is to have quite a long period of disconnection. In that case, ephemeral devices could get deleted from the admin console after a long enough period without connecting to control (this period may vary, but can be [as short as 30 minutes](https://tailscale.com/kb/1111/ephemeral-nodes?q=ephemeral#how-long-before-ephemeral-devices-are-auto-removed)). In that scenario, the pod will never recover on its own and it would definitely be good to make sure we can mark the pod as unhealthy, which I don't think our current health check will.

**Comment by chrismilson** [comment]
@tomhjp yes, the nodes are ephemeral nodes, and the disruption was due to a lengthy (>1 hour) ISP unplanned downtime window. Highly likely the nodes got deleted from the admin console during that time. Thanks for the tip :)

Since that seems to be my issue, is there anywhere I could start looking if I was interested in adding that as a feature to some kind of healthcheck?

**Comment by vvuk** [comment]
Also interested in this.

> In that scenario, the pod will never recover on its own and it would definitely be good to make sure we can mark the pod as unhealthy, which I don't think our current health check will.

I'm currently running into this issue -- we're deploying a container on a "serverless" GPU provider (RunPod) where a container is started, and then sits around idle/suspended(?) with the network disconnected until a request comes in from a higher level orchestrator. At that point the network is reconnected, but if we connected to our tailnet when the container is started, by this point the ephemeral node has been deleted and we just start getting "General SOCKS server failure" when trying to connect through the userspace SOCKS proxy.
